<h1>UK ROAD ACCIDENT DATA ANALYSIS</h1>

<h2>INCLUSIVE YEAR 2019 - 2022</h2>

<p>ANALYST: PETER PAUL F. ECLAVEA</p>

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import f_oneway

import warnings
warnings.filterwarnings('ignore')

In [2]:
accident = pd.read_csv('datasets//accident_data.csv')

In [3]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    05/06/2019  51.506187   
1       200701BS65737           Serious    02/07/2019  51.495029   
2       200701BS66127           Serious    26/08/2019  51.517715   
3       200701BS66128           Serious    16/08/2019  51.495478   
4       200701BS66837            Slight    03/09/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18/02/2022  57.374005   
660675  201091NM01881            Slight    21/02/2022  57.232273   
660676  201091NM01935            Slight    23/02/2022  57.585044   
660677  201091NM01964           Serious    23/02/2022  57.214898   
660678  201091NM02142           Serious    28/02/2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [4]:
accident[accident['District Area'] == 'Kensington and Chelsea']

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    05/06/2019  51.506187   
1       200701BS65737           Serious    02/07/2019  51.495029   
2       200701BS66127           Serious    26/08/2019  51.517715   
3       200701BS66128           Serious    16/08/2019  51.495478   
4       200701BS66837            Slight    03/09/2019  51.488576   
...               ...               ...           ...        ...   
533088  201001TB00796            Slight    14/11/2022  51.489306   
533316  201001TE00005           Serious    17/01/2022  51.520998   
533327  201001TE00016           Serious    02/02/2022  51.497193   
533337  201001TE00029           Serious    12/03/2022  51.480287   
533342  201001TE00038            Slight    24/03/2022  51.492639   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
533088        Darkness - lights lit  Kensington and Chelsea  -0.175749   
533316                     Daylight  Kensington and Chelsea  -0.213258   
533327                     Daylight  Kensington and Chelsea  -0.168230   
533337        Darkness - lights lit  Kensington and Chelsea  -0.185614   
533342        Darkness - lights lit  Kensington and Chelsea  -0.187716   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
533088                     2                   2             Wet or damp   
533316                     1                   1                     Dry   
533327                     1                   2             Wet or damp   
533337                     1                   1                     Dry   
533342                     1                   2             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
533088  Single carriageway               Urban     Fine no high winds   
533316  Single carriageway               Urban     Fine no high winds   
533327    Dual carriageway               Urban  Raining no high winds   
533337  Single carriageway               Urban     Fine no high winds   
533342          Roundabout               Urban  Raining no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
533088   Van / Goods 3.5 tonnes mgw or under  
533316                                   Car  
533327                                   Car  
533337                 Motorcycle over 500cc  
533

In [5]:
accident['Urban_or_Rural_Area'].value_counts()

Urban_or_Rural_Area
Urban          421663
Rural          238990
Unallocated        11
Name: count, dtype: int64

In [6]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [7]:
accident.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

<h1>Filling the missing data using fillna</h1>

In [8]:
accident['Latitude'] = accident['Latitude'].fillna('unknown')

In [9]:
accident['Longitude'] = accident['Longitude'].fillna('unknown')

In [10]:
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].fillna('unknown')

In [11]:
accident['Road_Type'] = accident['Road_Type'].fillna('unknown')

In [12]:
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].fillna('unknown')

In [13]:
accident['Weather_Conditions'] = accident['Weather_Conditions'].fillna('unknown')

<h1>Changing data types into proper ones</h1>

In [14]:
accident['Latitude'] = accident['Latitude'].astype('category')

In [15]:
accident['Longitude'] = accident['Longitude'].astype('category')

In [16]:
accident['Accident_Severity'] = accident['Accident_Severity'].astype('category')

In [17]:
accident['Light_Conditions'] = accident['Light_Conditions'].astype('category')

In [18]:
accident['District Area'] = accident['District Area'].astype('category')

In [19]:
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].astype('category')

In [20]:
accident['Road_Type'] = accident['Road_Type'].astype('category')

In [21]:
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].astype('category')

In [22]:
accident['Vehicle_Type'] = accident['Vehicle_Type'].astype('category')

In [23]:
accident['Weather_Conditions'] = accident['Weather_Conditions'].astype('category')

<h1>Converting Date Field into DATE TIME data type</h1>

In [24]:
accident['Accident Date'] = accident['Accident Date'].astype('str')
accident.dtypes
#remove spaces like 20 / 02 / 2025
accident['Accident Date'] = accident['Accident Date'].str.strip()
#replace / with -
accident['Accident Date'] = accident['Accident Date'].str.replace('/', '-')

In [25]:
accident['Accident Date'] = pd.to_datetime(accident['Accident Date'], dayfirst = True, errors = 'coerce')

<h1>Checking if there is still a null value</h1>

In [26]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

<h1>Adding new data columns</h1>

In [27]:
accident ['Year'] = accident['Accident Date'].dt.year
accident ['Month'] = accident['Accident Date'].dt.month
accident ['DayOfWeek'] = accident['Accident Date'].dt.dayofweek

In [28]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
Year                       0
Month                      0
DayOfWeek                  0
dtype: int64

<h1>QUESTION NO.1</h1>

<h2>Which year had the most accidents?</h2>

In [47]:
accident['Year'].value_counts()

Year
2019    182115
2020    170591
2021    163554
2022    144419
Name: count, dtype: int64

<h1>INSIGHT NO.1</h1>

<h2>The year 2019 had the most accidents covering a 182115 or 27.56% of 660679 total amount of accidents within the year 2019-2022.</h2>

<h1>QUESTION NO.2</h1>

<h2>What was the worst month by having the most accidents in 2019?</h2>

In [49]:
df_2019 = accident[accident['Year'] == 2019]
df_2019['Month'].count()

np.int64(182115)

<h1>INSIGHT NO.2</h1>

<h2>November had the highest number of accidents having a total number of 16559 or 9.09% of 182115.</h2>

<h1>QUESTION NO.3</h1>

<h2>What were the most common road surface conditions in November 2019</h2>

In [64]:
df_Nov2019 = df_2019[df_2019['Month'] == 11]
df_Nov2019['Road_Surface_Conditions'].value_counts()

Road_Surface_Conditions
Dry                     9458
Wet or damp             6786
Frost or ice             241
Snow                      28
Flood over 3cm. deep      23
unknown                   23
Name: count, dtype: int64

<h1>INSIGHT NO.3</h1>

<h2>There are a total of 9458 accidents occured in a dry surface road on November 2019</h2>

<h1>QUESTION NO.4</h1>

<h2>Which vehicle types were involved in accidents on a dry surface road on November 2019?</h2>

In [68]:
df_Nov2019[df_Nov2019['Road_Surface_Conditions'] == 'Dry']['Vehicle_Type'].value_counts()


Vehicle_Type
Car                                      6741
Bus or coach (17 or more pass seats)      595
Van / Goods 3.5 tonnes mgw or under       449
Motorcycle over 500cc                     426
Goods 7.5 tonnes mgw and over             293
Motorcycle 125cc and under                233
Taxi/Private hire car                     217
Motorcycle over 125cc and up to 500cc     128
Goods over 3.5t. and under 7.5t           109
Motorcycle 50cc and under                 102
Other vehicle                              94
Minibus (8 - 16 passenger seats)           41
Agricultural vehicle                       30
Data missing or out of range                0
Ridden horse                                0
Pedal cycle                                 0
Name: count, dtype: int64

<h1>INSIGHT NO.4</h1>

<h2>Cars were involved in most accidents probably due to it is the most common type of vehicle</h2>

<h1>QUESTION NO.5</h1>

<h2>Do more vehicles involved in an accident lead to more Casualties?</h2>

In [35]:
accident[["Number_of_Vehicles", "Number_of_Casualties"]].corr()

Number_of_Vehicles  Number_of_Casualties
Number_of_Vehicles              1.000000              0.228889
Number_of_Casualties            0.228889              1.000000

<h1>INSIGHT NO.5</h1>

<h2>More vehicles involved in an accident tend to result in more casualties.</h2>

<h1>QUESTION NO.6</h1>

<h2>What types of road have the most occuring fatal accidents</h2>

In [36]:
accident[accident["Accident_Severity"] == "Fatal"] ["Road_Type"].value_counts()

Road_Type
Single carriageway    6527
Dual carriageway      1815
Roundabout             142
One way street          95
Slip road               49
unknown                 33
Name: count, dtype: int64

<h1>INSIGHT NO.6</h1>

<h2>A single carriageway type of road is the most prone to fatal accidents</h2>

<h1>QUESTION NO.7</h1>

<h2>What type of vehicle is most likely to have an accident in a single carriageway with a Fatal accident.</h2>

In [69]:
accident[(accident["Road_Type"] == "Single carriageway") & 
           (accident["Accident_Severity"] == "Fatal")]["Vehicle_Type"].value_counts()


Vehicle_Type
Car                                      4973
Van / Goods 3.5 tonnes mgw or under       346
Bus or coach (17 or more pass seats)      249
Motorcycle over 500cc                     247
Goods 7.5 tonnes mgw and over             158
Motorcycle 125cc and under                150
Taxi/Private hire car                     114
Motorcycle over 125cc and up to 500cc      78
Motorcycle 50cc and under                  68
Other vehicle                              53
Goods over 3.5t. and under 7.5t            48
Minibus (8 - 16 passenger seats)           24
Agricultural vehicle                       14
Pedal cycle                                 5
Data missing or out of range                0
Ridden horse                                0
Name: count, dtype: int64

<h1>INSIGHT NO.7</h1>

<h2>A total of 4973 or 76.19% out of 6527 total Fatal accidents in a single carriageway involved cars</h2>

<h1>QUESTION NO.8</h1>

<h2>What month has the most occured accidents in 2020</h2>

In [37]:
df_2021 = accident[accident ['Year'] == 2021]

In [38]:
df_2021['Month'].sum()

np.int64(1093053)

In [39]:
df_2021['Month'].value_counts()

Month
11    15473
10    14834
7     14300
6     13936
5     13811
9     13792
12    13709
1     13417
8     13415
3     13202
4     12715
2     10950
Name: count, dtype: int64

<h1>INSIGHT NO.8</h1>

<h2>The most accident occured is the month of November which is almost 1.42% of the whole year of 2021</h2>

<h1>QUESTION NO.9</h1>

<h2>What kind of road surface condition is prone to accident in November 2021</h2>

In [40]:
df_Nov2021 = df_2021[df_2021['Month'] == 11]

In [41]:
df_Nov2021

Index Accident_Severity Accident Date   Latitude  \
353232  200901BS70618            Slight    2021-11-01  51.499011   
353239  200901BS70625            Slight    2021-11-01  51.497242   
353241  200901BS70627            Slight    2021-11-01  51.488898   
353242  200901BS70628            Slight    2021-11-03  51.482637   
353244  200901BS70630            Slight    2021-11-04  51.494521   
...               ...               ...           ...        ...   
516246       2.01E+12           Serious    2021-11-18  55.341751   
516248       2.01E+12            Slight    2021-11-30   55.04001   
516249       2.01E+12           Serious    2021-11-29  54.994538   
516251       2.01E+12            Slight    2021-11-27  55.063598   
516252       2.01E+12            Slight    2021-11-26  55.057373   

              Light_Conditions           District Area Longitude  \
353232                Daylight  Kensington and Chelsea -0.198558   
353239   Darkness - lights lit  Kensington and Chelsea -0.194593   
353241                Daylight  Kensington and Chelsea -0.190025   
353242                Daylight  Kensington and Chelsea -0.186385   
353244                Daylight  Kensington and Chelsea -0.158397   
...                        ...                     ...       ...   
516246  Darkness - no lighting   Dumfries and Galloway -3.490779   
516248                Daylight   Dumfries and Galloway -3.211317   
516249                Daylight   Dumfries and Galloway -3.067075   
516251  Darkness - no lighting   Dumfries and Galloway -3.400091   
516252  Darkness - no lighting   Dumfries and Galloway -3.065145   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
353232                     2                   2                     Dry   
353239                     1                   2                     Dry   
353241                     1                   1             Wet or damp   
353242                     1                   2             Wet or damp   
353244                     1                   2                     Dry   
...                      ...                 ...                     ...   
516246                     2                   4             Wet or damp   
516248                     1                   1                     Dry   
516249                     1                   1                     Dry   
516251                     1                   1             Wet or damp   
516252                     1                   1                     Dry   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
353232  Single carriageway               Urban      Fine + high winds   
353239          Roundabout               Urban     Fine no high winds   
353241      One way street               Urban  Raining no high winds   
353242  Single carriageway               Urban  Raining no high winds   
353244  Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
516246    Dual carriageway               Rural  Raining no high winds   
516248  Single carriageway               Rural     Fine no high winds   
516249  Single carriageway               Rural     Fine no high winds   
516251  Single carriageway               Rural     Fine no high winds   
516252  Single carriageway               Rural     Fine no high winds   

                               Vehicle_Type  Year  Month  DayOfWeek  
353232                                  Car  2021     11          0  
353239        Goods 7.5 tonnes mgw and over  2021     11          0  
353241  Van / Goods 3.5 tonnes mgw or under  2021     11          0  
353242  Van / Goods 3.5 tonnes mgw or under  2021     11          2  
353244                                  Car  2021     11          3  
...                                     ...   ...    ...        ...  
516246                Taxi/Private hire car  2021     11          3  
516248      Goods over 3.5t. and

In [42]:
df_Nov2021['Road_Surface_Conditions'].value_counts()

Road_Surface_Conditions
Wet or damp             9323
Dry                     5923
Frost or ice             111
Flood over 3cm. deep      84
unknown                   19
Snow                      13
Name: count, dtype: int64

<h1>INSIGHT NO.9</h1>

<h2>9,323 accidents, or 60.25%, were caused by wet roads.</h2>

<h1>QUESTION NO.10</h1>

<h2>What type of Vehicle is commonly involved in a wet road surface with fatal accidents on November 2021 aside from cars</h2>

In [71]:
df_Nov2021[(df_Nov2021["Road_Surface_Conditions"] == "Wet or damp") & 
           (df_Nov2021["Accident_Severity"] == "Fatal")]["Vehicle_Type"].value_counts()


Vehicle_Type
Car                                      78
Motorcycle over 500cc                     9
Van / Goods 3.5 tonnes mgw or under       6
Motorcycle 125cc and under                6
Goods 7.5 tonnes mgw and over             3
Taxi/Private hire car                     3
Motorcycle 50cc and under                 2
Bus or coach (17 or more pass seats)      2
Motorcycle over 125cc and up to 500cc     1
Goods over 3.5t. and under 7.5t           1
Minibus (8 - 16 passenger seats)          1
Data missing or out of range              0
Agricultural vehicle                      0
Other vehicle                             0
Ridden horse                              0
Pedal cycle                               0
Name: count, dtype: int64

<h1>INSIGHT NO.10</h1>

<h2>Motorcycles are the most commonly involved vehicles in fatal accidents on wet roads aside from cars, likely due to loss of control, skidding, and lack of protection.</h2>

<h1>QUESTION NO.11</h1>

<h2>Which vehicle type had the least number of accidents?</h2>

In [84]:
least_accident_vehicle = accident["Vehicle_Type"].value_counts().idxmin()
least_accident_vehicle_count = accident["Vehicle_Type"].value_counts().min()
print(f"Least involved vehicle type: {least_accident_vehicle} with {least_accident_vehicle_count} accidents")

accident["Vehicle_Type"].value_counts()

Least involved vehicle type: Ridden horse with 4 accidents


Vehicle_Type
Car                                      497992
Van / Goods 3.5 tonnes mgw or under       34160
Bus or coach (17 or more pass seats)      25878
Motorcycle over 500cc                     25657
Goods 7.5 tonnes mgw and over             17307
Motorcycle 125cc and under                15269
Taxi/Private hire car                     13294
Motorcycle over 125cc and up to 500cc      7656
Motorcycle 50cc and under                  7603
Goods over 3.5t. and under 7.5t            6096
Other vehicle                              5637
Minibus (8 - 16 passenger seats)           1976
Agricultural vehicle                       1947
Pedal cycle                                 197
Data missing or out of range                  6
Ridden horse                                  4
Name: count, dtype: int64

<h1>INSIGHT NO.11</h1>

<h2>Riding Horse is the least number of accident recorded within 2019-2022 with only 4 recorded accidents</h2>

<h1>QUESTION NO.12</h1>

<h2>Where do Ridden Horse accidents occured? is it only in Rural areas or both?</h2>

In [86]:
accident[accident["Vehicle_Type"] == least_accident_vehicle]["Urban_or_Rural_Area"].value_counts()


Urban_or_Rural_Area
Urban          3
Rural          1
Unallocated    0
unknown        0
Name: count, dtype: int64

<h1>INSIGHT NO.12</h1>

<h2>3 of them are occured in an Urban area while the other 1 occured in a Rural Area.</h2>

<h1>QUESTION NO.13</h1>

<h2>What Day of the Week are these accidents commonly occurs?</h2>

In [90]:
accident[accident["Vehicle_Type"] == least_accident_vehicle]["DayOfWeek"].value_counts()

DayOfWeek
3    2
1    1
2    1
Name: count, dtype: int64

<h1>INSIGHT NO.13</h1>

<h2>All accidents of the Ridden Horse occur in weekdays, 2 in Thursday and 1 in both Tuesday and Wednesday.</h2>

<h1>QUESTION NO.14</h1>

<h2></h2>